PocketKit/Sources/Analytics/Events/Engagement.swift

```diff
@@ -0,0 +1,75 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import Foundation
+import class SnowplowTracker.SelfDescribing
+
+/**
+ * Event triggered when a user engages with a UI element.
+ */
+public struct Engagement: Event {
+    public static let schema = "iglu:com.pocket/engagement/jsonschema/1-0-1"
+
+    /**
+     * Indicates the type of engagement.
+     */
+    let type: Engagement.`Type`
+
+    /**
+     * The new value of a setting/filter, if the user engaged with something and modified its state in doing so.
+     */
+    let value: String?
+
+    let uiEntity: UiEntity
+
+    let extraEntities: [Entity]
+
+    public init(_ type: Engagement.`Type` = .general, value: String? = nil, uiEntity: UiEntity, extraEntities: [Entity] = []) {
+        self.type = type
+        self.value = value
+        self.uiEntity = uiEntity
+        self.extraEntities = extraEntities
+    }
+
+    public func toSelfDescribing() -> SelfDescribing {
+        var payload: [String: NSObject] = [
+            "type": NSString(string: type.toType().value),
+        ]
+
+        if let value = value {
+            payload["value"] = NSString(string: value)
+        }
+
+        let base = SelfDescribing(schema: Engagement.schema, payload: payload)
+        base.contexts.add(uiEntity.toSelfDescribingJson())
+        extraEntities.forEach { base.contexts.add($0.toSelfDescribingJson()) }
+        type.toType().requiredEntities.forEach { base.contexts.add($0.toSelfDescribingJson()) }
+
+        return base
+    }
+
+    public enum `Type` {
```

Comment:
is it strictly needed to name this type `Type`? Would it be ok to rename it to `EngagementType` or something to avoid confusion and proliferation of ticks?